ui/src/features/stage/freight-history.tsx

```diff
@@ -0,0 +1,164 @@
+import { createConnectQueryKey } from '@connectrpc/connect-query';
+import { faHistory } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { useQueryClient } from '@tanstack/react-query';
+import { Empty } from 'antd';
+import classNames from 'classnames';
+import { useMemo } from 'react';
+import { generatePath, Link, useNavigate } from 'react-router-dom';
+
+import { paths } from '@ui/config/paths';
+import freightTimelineStyles from '@ui/features/freight-timeline/freight-timeline.module.less';
+import { queryFreight } from '@ui/gen/service/v1alpha1/service-KargoService_connectquery';
+import { QueryFreightResponse } from '@ui/gen/service/v1alpha1/service_pb';
+import {
+  Freight,
+  FreightReference,
+  FreightRequest,
+  StageStatus
+} from '@ui/gen/v1alpha1/generated_pb';
+
+import { FreightContents } from '../freight-timeline/freight-contents';
+import { FreightItemLabel } from '../freight-timeline/freight-item-label';
+import { FreightTimelineWrapper } from '../freight-timeline/freight-timeline-wrapper';
+
+import requestedFreightStyles from './requested-freight.module.less';
+
+export const FreightHistory = ({
+  projectName,
+  freightHistory,
+  requestedFreights,
+  className
+}: {
+  className?: string;
+  requestedFreights: FreightRequest[];
+  projectName: string;
+  // show the freight history thats 1:1 with requested freight
+  freightHistory?: StageStatus['freightHistory'];
+  // freight hash name which is active at the moment
+  // you can get this from lastPromotion in stage status
+  // usually last one is active but we have to consider multi-pipeline case
+  currentActiveFreight?: string;
+}) => {
+  const queryClient = useQueryClient();
+  const navigate = useNavigate();
+
+  const { freightHistoryPerWarehouse, freightMap } = useMemo(() => {
+    // to show the history
+    const freightHistoryPerWarehouse: Record<
+      string /* warehouse eg. Warehouse/w-1 or Warehouse/w-2 */,
+      FreightReference[]
+    > = {};
+
+    for (const freightCollection of freightHistory || []) {
+      // key - value
+      // warehouse identifier - freight reference
+      const items = freightCollection?.items || {};
+
+      for (const [warehouseIdentifier, freightReference] of Object.entries(items)) {
+        if (!freightHistoryPerWarehouse[warehouseIdentifier]) {
+          freightHistoryPerWarehouse[warehouseIdentifier] = [];
+        }
+
+        freightHistoryPerWarehouse[warehouseIdentifier].push(freightReference);
+      }
+    }
+
+    const freightData = queryClient.getQueryData(
```

Comment:
> Why not regular fetch? I think you assumed there are data. Maybe it will work always in our case, but not sure if it is a good practice. You can set a stale property for data in query configuration, so it won't generate a new request. Also, you are sure that if the data are not available, it will be fetched. Let me know what you think.

You are right, this will work in our case. I didn't want to make the query for the data we already have and I wanted to avoid adding context provider or prop drilling to have that data. The closest I can do is [ensureQueryData](https://tanstack.com/query/v4/docs/reference/QueryClient/#queryclientensurequerydata) but since its async I have to wrap in effect. I will look whether to make query with invalidation or use ensureQueryData and add state